### ISSUE TITLE
About the speed at which pseudo-tags are generated

### ISSUE THREAD

**Issue by yumu-173** [facebookresearch/CutLER#74]
When I use the maskcut.py to generate pseudo-labels, it takes 3 hours for each set of data in ImageNet, and there are 1000 sets of such data, is this normal, and what can I do to improve the running speed
Hope to get your answer, I will be very grateful,

```
Loading weight from https://dl.fbaipublicfiles.com/dino/dino_vitbase8_pretrain/dino_vitbase8_pretrain.pth
Load base pre-trained feature...
  0%|▊                                                                                                                                                                              | 6/1300 [01:22<4:51:28, 13.52s/it
```

**Comment by TJ12342** [comment]
I've found a good way.
Replace function second_smallest_eigenvector with the following function. (If you have a GPU)

`def second_smallest_eigenvector_torch(A_np, D_diag_np, device='cuda'):
    import numpy as np, torch

    Dsi = torch.from_numpy((1.0 / (np.sqrt(D_diag_np) + 1e-12)).astype(np.float32)).to(device)
    A   = torch.from_numpy(A_np.astype(np.float32)).to(device).contiguous()

    if A.dim() != 2:
        A = A.squeeze()
    n = A.shape[0]

    A.fill_diagonal_(0.0)
    A = 0.5 * (A + A.T)
    A = A * Dsi  
    A = (A.t() * Dsi).t()
    A = -A
    A.diagonal().add_(1.0) 
    vals, vecs = torch.linalg.eigh(A)   # ascending
    v2 = vecs[:, 1].detach().cpu().numpy()
    return v2.copy(), v2
`
